neps/nep-0509.md

```diff
@@ -0,0 +1,231 @@
+* State sync is enabled (so that nodes can track different shards across epochs)
+* Merkle Patricia Trie continues to be the state trie implementation
+* TBD
+
+### High level requirements
+
+* No validator needs to track all shards.
+* Security of protocol must not degrade.
+  * Validator assignment for both chunk validation and block validation should not create any security vulnerabilities.
+* Block processing time should not take significantly more than what it takes today.
+* Any additional load on network and compute should not negatively affect existing functionalities of any node in the blockchain.
+  * The cost of additional network and compute should be acceptable.
+* Validator rewards should not be reduced.
+* Resharding should still be possible after stateless validation is in place.
+* TBD
+
+### Out of scope
+
+* Data size optimizations such as compression, for both chunk data and state witnesses, except basic optimizations that are practically necessary.
+* Separation of consensus and execution, where consensus runs independently from execution, and validators asynchronously perform state transitions after the transactions are proposed on the consensus layer, for the purpose of amortizing the computation and network transfer time.
+* More shards - this is covered in the resharding project.
+* ZK integration.
+* Underlying data structure change (e.g. verkle tree).
+* Change to validator rewards.
+* TBD
+
+## High level flow
+
+We propose a change to the following parts of the chunk and block production flow:
+
+* When a chunk producer produces a chunk, in addition to collecting transactions and receipts for the chunk, it will also produce a `ChunkStateWitness`.
+  * The `ChunkStateWitness` contains whatever data necessary to prove that this chunk's header should indeed be what is being produced:
+    * As it is today, all fields of the `ShardChunkHeaderInnerV2`, except `tx_root`, are uniquely determined by the blockchain's history based on where the chunk is located (i.e. its parent block and shard ID).
+    * The `tx_root` is based on the list of transactions proposed, which is at the discretion of the chunk producer. However, these transactions must be valid (i.e. the sender accounts have enough balance and the correct nonce, etc.).
+    * This `ChunkStateWitness` proves to anyone, including those who track only block data and no shards, that this chunk header is correct, meaning that the uniquely determined fields are exactly what should be expected, and the discretionary `tx_root` field corresponds to a valid set of transactions.
+  * The `ChunkStateWitness` is not part of the chunk itself; it is distributed separately and is considered transient data.
+* The chunk producer then distributes the `ChunkStateWitness` to a subset of *Chunk Validators* assigned for this shard. This is in addition to, and independent of, the existing chunk distribution logic (implemented by `ShardsManager`) today.
```

Comment:
Should we talk about chunk witness compression, size limit, and reed-solomon encoding?

cc. @shreyan-gupta, @jancionear , @pugachAG

Reply:
please check below comment as well

Reply:
@walnut-the-cat I wrote some documentation on the size limits, could be added to the NEP: https://github.com/near/nearcore/pull/11559